backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/AbstractInvoiceDAO.java

```diff
@@ -413,4 +420,61 @@ public boolean hasCompletedInvoicesReferencing(@NonNull final InvoiceId invoiceI
 				.create()
 				.list();
 	}
+
+	public Optional<InvoiceId> retrieveIdByInvoiceQuery(InvoiceQuery query)
+	{
+		if (query.getInvoiceId() != null)
+		{
+			return Optional.of(InvoiceId.ofRepoId(query.getInvoiceId()));
+		}
+		if (query.getExternalId() != null)
+		{
+			return getInvoiceIdByExternalIdIfExists(query);
+		}
+		if (!Check.isEmpty(query.getDocType()))
+		{
+			return getInvoiceIdByDocumentIdIfExists(query);
+		}
+		return Optional.empty();
+	}
+
+	private Optional<InvoiceId> getInvoiceIdByDocumentIdIfExists(@NonNull final InvoiceQuery query)
+	{
+		final String documentNo = assumeNotNull(query.getDocumentNo(), "Param query needs to have a non-null docId; query={}", query);
+		final OrgId orgId = assumeNotNull(query.getOrgId(), "Param query needs to have a non-null orgId; query={}", query);
+		final DocBaseAndSubType docType = assumeNotNull(query.getDocType(), "Param query needs to have a non-null docType; query={}", query);
```

Comment:
Same remarks as with the orderDAO (=>also use DocSubType and fail if more than one)

Reply:
added DocSubType and logic to use it.